### DISCUSSION TITLE
Unknown Status of 2 From Network

### DISCUSSION THREAD

**Discussion by DS256** [micropython/micropython#10746] [RP2040 / Pico]
I'm getting a status of 2 coming back from WLAN.status() after WLAN.connect(...). What does the 2 mean? I can't see it listed in help(network)
Here's the code snippet
            wlan.connect(secrets.ssid, secrets.pwd)
            sleep(3)
            # If the status is not 'connecting' (1) or connected abort
            status = wlan.status() 
            if not ((status == 1) or wlan.isconnected() ):

This is what I see from help(network)
  STAT_IDLE -- 0
  STAT_CONNECTING -- 1
  STAT_WRONG_PASSWORD -- -3
  STAT_NO_AP_FOUND -- -2
  STAT_CONNECT_FAIL -- -1
  STAT_GOT_IP -- 3

**Comment by davefes** [comment]
My understanding is that Micropython error messages are or were preceded by a - sign to distinguish them from "normal" Python errors.

Hmmm, I see some have - signs and another ones don't.

Can you do a scan?

I use this on my ESPNow systems and it should give you more information:

https://github.com/glenn20/micropython-espnow-utils  
try wifi.py

**Reply by DS256** [reply]
I have no other issue other than if I check status, sometimes it's 2 which is not defined by any of the STAT_* values.

**Reply by th3w** [reply]
Status 2 could be STAT_NOIP, reading the LINK Status definitions directly in the cyw43 driver [here](https://github.com/georgerobotics/cyw43-driver/blob/9bfca61173a94432839cd39210f1d1afdf602c42/src/cyw43.h#L82) in line 80-86

**Reply by DS256** [reply]
Thanks @th3w. That makes sense.

**Comment by pinqwing** [comment]
Me too gets the status() == 2 after an apperently successful connect.

'connected:
  as PicoW 
  at reticulum novum
signal strength: good
connection.status = 2
`
the info coming  from  connection.config(...), but indeed the ifconfig() does not look correct:

('0.0.0.0', '255.255.255.0', '192.168.0.1', '8.8.8.8')


`for a in dir (network):
    if a.startswith("STAT"):
        print(a, "=", getattr(network, a))

STAT_CONNECTING = 1
STAT_CONNECT_FAIL = -1
STAT_GOT_IP = 3
STAT_IDLE = 0
STAT_NO_AP_FOUND = -2
STAT_WRONG_PASSWORD = -3
`

**Reply by KnorkFox** [reply]
It looks like, you got no ip adress from the WLAN.  it must be a private ipV4 adress.
This should look like 192.168.0.11, 255.255.255.0, 192.168.0.1, 8.8.8.8
The first one should be your adress, then the netmask and the server adress.

**Reply by pinqwing** [reply]
No, my code and hardware work. Mostly, that is. I now let it do a few retries until I get the GOT_IP, and up to now (not very long) that works. I cannot say I completely understand it, but it seems to be related to cold starting the pico. When connected to the USB and run through Thonnie, making the connection is so much easier, that I suspect either the controller or the router does not completely disconnect. I have the impression that after a cold start, the first connection attempt returns -1 or -3, the second +2, and the subsequent +3. The controller's mac-address has a fixed IP leased from the router, so it may be the router that is to reluctant to make the ip-address available again after the controller was disconnected. Another possibility is that more time is required after leaving the while status() == CONNECTING loop and testing the new status. I now have a sleep(2) to add some time. When I have a bit more time, I'll rewrite the module in Asyncio-style.

**Reply by Josverl** [reply]
For the PICO_W using the CYW43 the link status 2 refers to :

[`#define CYW43_LINK_NOIP         (2)     ///< Connected to wifi, but no IP address`](https://github.com/georgerobotics/cyw43-driver/blob/dd7568229f3bf7a37737b9e1ef250c26efe75b23/src/cyw43.h#L99-L105).

As you found you need to wait 'some time' to allow for the router/dhcp server to check its pool, assignment and allow for the offer to be accepted.

perhaps also useful: a simple [explanation of DHCP](https://www.computernetworkingnotes.com/ccna-study-guide/how-dhcp-works-explained-with-examples.html)

**Reply by pinqwing** [reply]
Thanks, that was very informative. It suggests that just waiting a few seconds after receiving NOIP (2) the queried status just might change into GOT_IP/UP (3).  I would have tried that just now, but unfortunately today (with another picoW-board) I am not getting the NOIP again. Either way, my workaround to repeat the connection request also overcomes the steady first connection failure after a cold boot.
